### PR TITLE
Sk0 96 invalid email address is handled

### PR DESCRIPTION
**Pull request by Thomas851204**
On the FE the Validators.email formcontrol passed through 2 edge cases:

- the domain is missing (eg.: test@test). In this case the backend bounced the email back with 400 statuscode, so the user wasnt created.

- the name of the email had special characters (eg.: tes&t@test.test). In this case both FE and BE let it through and the user was created

Needed to create a custom validator to handle these cases.

### CODE REVIEW

**Comment by Thomas851204** [comment]
The "Weak password" e2e scenario still doesnt work, need to resolve.

File: e2e/features/steps/registration.steps.ts [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+});
+
+When("the user clicks the {string} button", async function (register: string) {
+  await page.getByLabel(register).click();
+});
+
+Then(
+  "the system should display an error message indicating the email address is invalid",
+  async function () {
+    let alert = page.getByRole("alertdialog");
```
Comment (markkovari) [inline e2e/features/steps/registration.steps.ts]:
```suggestion
    const alert = page.getByRole("alertdialog");
```

Might be const if no redeclaration.

**Comment by markkovari** [comment]
Well honestly (retrospectively) I definitely should've added a test-run description how to run the the tests. Without it we can just guess or figure it out ourself. (which shouldn't be the case)

mine was/is in the root folder

```bash
docker compose up --force-recreate --remove-orphans --build --detach

cd e2e

npm i

npm t
```

we should unify this one too, since I cannot figure out how you run yours.

File: e2e/features/registration.feature [hunk 2]
```diff
@@ -1,10 +1,33 @@
 Feature: Registration
 
-  Scenario: Weak Password
+
+  Scenario Outline: Invalid email address
     Given the user is on the registration page
+    When the user enters an invalid "<emailAddress>" address (e.g., missing @)
```
Comment (markkovari) [inline e2e/features/registration.feature]: The scenario outline should not contain specifics/examples in the steps, you should rather have this as an input (as you do too).

**Review by markkovari — CHANGES_REQUESTED** [review]

File: sk0-front/src/app/common/services/registerEmailValidator.ts [hunk 3]
```diff
@@ -0,0 +1,26 @@
+import { AbstractControl, ValidationErrors, ValidatorFn } from "@angular/forms";
+import { Observable, of } from "rxjs";
```
Comment (markkovari) [inline sk0-front/src/app/common/services/registerEmailValidator.ts]: This import is not used (as you can see it in the sonarscanner output), you can delete it.

File: sk0-front/src/app/common/services/registerEmailValidator.ts [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+
+    const nameNoSpecChar = /^[a-zA-Z0-9._%+-]+$/.test(username);
+
+    const domainExists = !!domain;
+
+    const emailValid = nameNoSpecChar && domainExists;
+
+    return !emailValid ? { invalidEmail: true } : null;
+  };
+}
```
Comment (markkovari) [inline sk0-front/src/app/common/services/registerEmailValidator.ts]: This file is not really a service, I think you might wanted to create another package with `utils/validators` or what the name that you would like. And probably you should unit test it, probably we might wanted to use this function at different places as well with different context.

File: sk0-front/src/app/auth/register/register.page.ts [hunk 5]
```diff
@@ -59,6 +64,14 @@ export class RegisterPage {
       );
       return;
     }
+
+    if (this.registerForm.get("email")?.invalid) {
```
Comment (markkovari) [inline sk0-front/src/app/auth/register/register.page.ts]: We have this line or field which is similar (`this.registerForm.get("email")`, we probably might wanted to simplify this part.

**Review by markkovari — CHANGES_REQUESTED** [review]

File: sk0-front/src/app/common/services/registerEmailValidator.ts [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+      return null;
+    }
+
+    const [username, serverDomain] = control.value.split("@");
+
+    if (!username || !serverDomain) {
+      return null;
+    }
+
+    const [server, domain] = serverDomain.split(".");
```
Comment (markkovari) [inline sk0-front/src/app/common/services/registerEmailValidator.ts]:
```suggestion
    const [, domain] = serverDomain.split(".");
```
if the first element is not used, you can omit it like this or prefix with an `_`, by convention.

```suggestion
    const [_server, domain] = serverDomain.split(".");
```

File: e2e/features/registration.feature (hunk 2 above)
Comment (Thomas851204) [reply e2e/features/registration.feature]: Im not sure what you mean. In the invalid email scenario the emailaddress is an input at the "when" step, as thats what Im testing and have examples of.

In the weak password scenario you also have the password as an input, so what should be changed here? 

Or do you mean that ALL input fields should be treated as an input in both scenarios and specify the constant values in the steps file as const variables?

File: sk0-front/src/app/common/services/registerEmailValidator.ts (hunk 4 above)
Comment (Thomas851204) [reply sk0-front/src/app/common/services/registerEmailValidator.ts]: These custom validators are basically just a single function, I could've put it into the component where Im actually implementing it (register.page.ts), but that didnt seem right, thats why I chose to put it in the common/service folder.
You are right, its not a service, so then I will move it to common/utils.

File: sk0-front/src/app/auth/register/register.page.ts (hunk 5 above)
Comment (Thomas851204) [reply sk0-front/src/app/auth/register/register.page.ts]: I dont really see how. If I say at line 57 that
const email = this.registerForm.get("email");
then I can rewrite line 68 in a simpler form (also line 61), but there will be a problem for the store.dispatch command, as it expects a string type, and email.value would be an AbstractControl<string | null, string | null> type.
So, from there I would have to rewrite the email's type in the store action, effect and reducer.

I dont really know how  I could resolve 'const email = this.registerForm.get("email")' into a string, as email?.value doesnt seem to work in the store registerStarted dispatch...

**Comment by Thomas851204** [comment]
> Well honestly (retrospectively) I definitely should've added a test-run description how to run the the tests. Without it we can just guess or figure it out ourself. (which shouldn't be the case)
> 
> mine was/is in the root folder
> 
> ```shell
> docker compose up --force-recreate --remove-orphans --build --detach
> 
> cd e2e
> 
> npm i
> 
> npm t
> ```
> 
> we should unify this one too, since I cannot figure out how you run yours.

I start my frontend with ng serve (so, port 4200), start backend (though not really necessary for this) with nest start --watch, then navigate to e2e and run the tests with npm t.
I looked at the package.json of the frontend but I couldnt find a script that would launch the app on the 8080 port (but I might have overlooked something).
Additionally in const.ts in the e2e folder, we have this line:
const frontendPort = process.env.FRONTEND_PORT || 8080;
Since process.env also specifies port 8080, the only way I can test my setup if i do it via CLI (ng serve --port), or changing my process.env.

What is your suggestion for unification, because right now I can either test only my scenario on 4200 or yours on 8080, but not both.

File: e2e/features/registration.feature (hunk 2 above)
Comment (markkovari) [reply e2e/features/registration.feature]:
```suggestion
    When the user enters an invalid "<emailAddress>" address
```

I meant that the example at the end of the sentence (or "Step") is one of the examples, I would rather not create a matcher with it, since it is in the examples "array".

**Comment by markkovari** [comment]
Well, good question. I am not running the application in watch mode, when running the end-to-end test cases, the apps should not change on file-saves during the e2e test runs. We can definitely create two different script as well to run the tests against different urls in the scripts of the package.json.

Not sure if you have been looking into #88, you can see that "my scripts are working" remotely too. Not saying yours wont, but we should have a setup, which will enable agents (users and other machines too), to run the scripts out of the box.

And the easiest way is probably to add an npm scripts with a prefix. And I do not mind to fix mine, if you wanted to use the `npm t`.

File: sk0-front/src/app/auth/register/register.page.ts (hunk 5 above)
Comment (markkovari) [reply sk0-front/src/app/auth/register/register.page.ts]: I meant something like this, you do not need to use this.

```ts
const emailControl = this.registerForm.get("email");
const email = emailControl?.value;
const password = this.registerForm.get("password")?.value;
const name = this.registerForm.get("name")?.value;

if (!email || !password || !name) {
  await this.notificationService.alertError(
    "Kérlek tölts ki minden mezőt!"
  );
  return;
}

if (emailControl?.invalid) {
  await this.notificationService.alertError(
    "Kérlek adj meg egy érvényes email címet!"
  );
  return;
}

```

**Comment by markkovari** [comment]
Yep this is looking good too:
```ts
{
  "name": "e2e",
  "version": "1.0.0",
  "description": "End-to-end testing with Cucumber/Playwright",
  "main": "index.js",
  "scripts": {
    "test:prod": "cucumber-js",
    "test": "FRONTEND_PORT=4200 cucumber-js"
  },
  "author": "",
  "license": "ISC",
  "devDependencies": {
    "@cucumber/cucumber": "^10.3.2",
    "@playwright/test": "^1.42.1",
    "@types/node": "^20.12.4",
    "ts-node": "^10.9.2",
    "typescript": "^5.4.3"
  },
  "dependencies": {
    "dotenv": "^16.4.5"
  }
}


```

File: sk0-front/src/app/auth/register/register.page.ts (hunk 5 above)
Comment (Thomas851204) [reply sk0-front/src/app/auth/register/register.page.ts]: Okay, got it. Fixed.

**Review by markkovari — APPROVED** [review]

**Review by rolcho — APPROVED** [review]

**Comment by Thomas851204** [comment]
Made a small mistake. Didnt merge here, but instead pulled changes from origin dev into my local branch, handled conflicts there in vscode, and committed/pushed the changes back to remote, so I'd need the two approvals again...

**Review by markkovari — APPROVED** [review]
